File: code_snippets/python/src/use_cases/microservices/idempotency.py
import requests

from src.use_cases.microservices.subscription_service import SubscriptionRequest

RESTATE_URL = "http://localhost:8080"
subscription_request = SubscriptionRequest(user_id="user123", credit_card="1234", subscriptions=["sub1", "sub2"])
request_id = "myTask123"
# <start_here>
# <mark_1>
response = requests.post(
    url = f"${RESTATE_URL}/SubscriptionService/add",
    json=subscription_request,
    headers={
        # <mark_2>
        "idempotency-key": request_id,
        # </mark_2>
        "Content-Type": "application/json"
    }
)
# </mark_1>
# <end_here>

File: code_snippets/ts/src/use_cases/microservices/idempotency.ts
import * as restate from "@restatedev/restate-sdk-clients";
import { Opts } from "@restatedev/restate-sdk-clients";
import express from "express";
import { SubscriptionService } from "./subscription_service";
import { randomUUID } from "node:crypto";

const app = express();

const RESTATE_URL = "localhost:8080";

app.get("/reserve/:product/:reservationId", async (req, res) => {
  const requestId = randomUUID().toString();
  const subscriptionRequest = {
    userId: "123",
    creditCard: "1234-1234-1234",
    subscriptions: ["Netflix", "Disney+"],
  };
  // <start_here>
  // <mark_1>
  const restateClient = restate.connect({ url: RESTATE_URL });
  const reservation = await restateClient
    .serviceClient<SubscriptionService>({ name: "SubscriptionService" })
    .add(
      // target handler
      subscriptionRequest,
      // <mark_2>
      Opts.from({ idempotencyKey: requestId })
      // </mark_2>
    );
  // </mark_1>
  // <end_here>

  res.json(reservation);
});
